src/thread_pool.cpp

```diff
@@ -75,34 +77,51 @@ auto thread_pool::shutdown() noexcept -> void
     }
 }
 
-auto thread_pool::executor(std::stop_token stop_token, std::size_t idx) -> void
+auto thread_pool::executor(std::size_t idx) -> void
 {
     if (m_opts.on_thread_start_functor != nullptr)
     {
         m_opts.on_thread_start_functor(idx);
     }
 
-    while (!stop_token.stop_requested())
+    auto drain_queue = [this](std::unique_lock<std::mutex>& lk) -> void
     {
-        // Wait until the queue has operations to execute or shutdown has been requested.
-        while (true)
+        // Process this batch until the queue is empty.
+        while (!m_queue.empty())
         {
-            std::unique_lock<std::mutex> lk{m_wait_mutex};
-            m_wait_cv.wait(lk, stop_token, [this] { return !m_queue.empty(); });
-            if (m_queue.empty())
-            {
-                lk.unlock(); // would happen on scope destruction, but being explicit/faster(?)
-                break;
-            }
-
             auto handle = m_queue.front();
             m_queue.pop_front();
 
-            lk.unlock(); // Not needed for processing the coroutine.
-
+            // Release the lock while executing the coroutine.
+            lk.unlock();
             handle.resume();
             m_size.fetch_sub(1, std::memory_order::release);
+            lk.lock();
         }
+    };
+
+    // Process until shutdown is requested and the total number of tasks reaches zero.
+    while (!m_shutdown_requested.load(std::memory_order::acquire))
```

Comment:
todo: combine these, I don't think the `wait_for` on the cv is required if we find the bug